internal/unionstore/memdb_arena.go

```diff
@@ -331,6 +337,47 @@ func (l *memdbVlog) getValue(addr memdbArenaAddr) []byte {
 	return block[valueOff:lenOff:lenOff]
 }
 
+// Get value, and possibly write a new vlog entry in the latest stage to let the locking phase be aware of it via InspectStage.
+// It should be used where the read result affect users or upper level modules. Temporary flags need to be unset in such cases.
+func (l *memdbVlog) getValue(addr memdbArenaAddr) (value []byte) {
+	if !l.memdb.enableTemporaryFlags {
+		return l.pureGetValue(addr)
+	}
+	hdrOffset := addr.off - memdbVlogHdrSize
+	block := l.blocks[addr.idx].buf
+	var hdr memdbVlogHdr
+	hdr.load(block[hdrOffset:])
+
+	if hdr.valueLen == 0 {
+		value = tombstone
+	} else {
+		valueOff := hdrOffset - hdr.valueLen
+		value = block[valueOff:hdrOffset:hdrOffset]
+	}
+
+	l.processFlagNeedConstraintCheckInPrewrite(addr, hdr.nodeAddr, value)
+
+	return value
+}
+
+// remove the temporary flag NeedConstraintCheckInPrewrite
+func (l *memdbVlog) processFlagNeedConstraintCheckInPrewrite(valueAddr memdbArenaAddr, nodeAddr memdbArenaAddr, value []byte) {
+	node := l.memdb.getNode(nodeAddr)
+	flags := node.getKeyFlags()
+	// only process if current access is on the latest version, so that we do it at most once.
+	if flags.HasNeedConstraintCheckInPrewrite() && node.vptr == valueAddr {
+		flags = kv.ApplyFlagsOps(flags, kv.DelNeedConstraintCheckInPrewrite)
+		node.setKeyFlags(flags)
+
+		// if this is not in the latest stage, we need to copy the vlog entry to the latest stage, marking it as
+		// a modification (of flag) so that the locking phase could find it via `InspectStage` and acquire the lock.
+		latestStagingHandle := len(l.memdb.stages)
+		if latestStagingHandle > 0 && !valueAddr.AfterCheckPoint(l.memdb.stages[latestStagingHandle-1]) {
+			l.memdb.setValue(node, value)
```

Comment:
Considering the worst case, if the user mistakenly enables the lazy check and executes select statements after every insert statement, would it be possible there could be 2x more memory usage?
Maybe we need add adapt this change to the previous memory tracking enhancements so it could be diagnosed.

Reply:
I think so, but it would be less than 2x. The memory tracker already tracks the value log. 
Did you mean separating the value log space used for such rewrite? That would be indirect since we can only track the  sizes of values. And there is no way to let users query the size, for now.